File: c++/leetcode/496. Next Greater Element I.cpp
class Solution
{
public:
  vector<int> nextGreaterElement(vector<int> &nums1, vector<int> &nums2)
  {
    unordered_map<int, int> seen;
    stack<int> stk;

    for (auto i : nums2)
    {
      while (stk.size() && stk.top() < i)
      {
        seen[stk.top()] = i;
        stk.pop();
      }

      stk.push(i);
    }

    for (int i = 0; i < nums1.size(); ++i)
      nums1[i] = seen.count(nums1[i]) ? seen[nums1[i]] : -1;

    return nums1;
  }
};

File: javascript/leetcode/496. Next Greater Element I.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function (nums1, nums2) {
  const seen = new Map();
  const stk = [];

  for (let i of nums2) {
    while (stk.length && stk[stk.length - 1] < i) seen.set(stk.pop(), i);

    stk.push(i);
  }

  console.log(seen);

  for (let i = 0; i < nums1.length; ++i)
    nums1[i] = seen.has(nums1[i]) ? seen.get(nums1[i]) : -1;

  return nums1;
};
